cloud_storage: Fix race condition in topic recovery

Lazin opened the pull request:
## Cover letter

This PR fixes race condition in topic recovery.

If during topic recovery any partition 0 replica will complete recovery before any other replica will start downloading the data it will read updated `topic_manifest.json`. This manifest will have revision of the newly created topic instead of the revision of the old one (the one which is being recovered). 

The problem is fixed by downloading `topic_manifest.json` only once and passing parameters needed for recovery (and extracted from topic manifest) to every replica of every partition. The mechanics here is similar to the one used by read replicas. The `topic_manifest.json` is downloaded in `topics_frontend` using the same component which read replicas are using. The data needed by topic recovery is stored in `remote_topic_properties` in `topic_configuration` which is also used by read replicas.
This is OK because it doesn't make sense to use topic recovery on read replica.

Fixes: https://github.com/redpanda-data/redpanda/issues/5474

## UX changes

NA

## Release notes

* Fix race condition in topic recovery

### Features

NA

### Improvements

* Fix race condition in topic recovery

dotnwat commented on src/v/cloud_storage/topic_manifest.cc:
```diff
@@ -154,6 +157,7 @@ struct topic_manifest_handler
     std::optional<int32_t> _partition_count;
     std::optional<int16_t> _replication_factor;
     std::optional<model::initial_revision_id> _revision_id{};
+    std::optional<model::initial_revision_id> _prev_revision_id{};
```
nit: just wondering what purpose using a std::optional has here. from what i can tell any time the revision doesn't have a value its given a value of ::min() so it could be initialized directly to ::min?

Lazin replied:
This would definitely work. But I just decided to follow current approach. Otherwise my field would be out of place, and the reader of the code will definitely have questions like "why everything is optional except this field".

dotnwat commented on src/v/cloud_storage/partition_recovery_manager.cc:
```diff
@@ -292,16 +292,15 @@ partition_downloader::download_log(const remote_manifest_path& manifest_key) {
+        vlog(
+          _ctxlog.warn,
+          "Can't upload partition manifest {} after recovery",
+          target.get_manifest_path());
```
based on what is happening below it sort of looks like the intention here was to return a `log_recovery_result` rather than also checking if the manifest revision is not the initial revision and potentially changing it. at least returning here would be a big clearer and looks like it would be correct? i guess it all hinges on:

```
    if (mat.topic_manifest.get_revision() != _ntpc.get_initial_revision()) {
        mat.topic_manifest.set_revision(_ntpc.get_initial_revision());
```

ajfabbri replied:
Are there any issues with failling to upload one manifest, but not the other? I.e. we fail to upload partition manifest but succeed in uploading topic manifest.

Lazin commented:
The failures are unrelated to topic recovery. I ran it with `ci-repeat-5` and all topic recovery tests including the ones which was marked as `ok_to_fail` passed.

ajfabbri commented on tools/cloud_storage_manifest_patcher/main.py:
```diff
@@ -0,0 +1,393 @@
+    pass
+
+
+S3ObjectMetadata = namedtuple('S3ObjectMetadata',
```
Typed version of namedtuple is an improvement. Example [here](https://github.com/redpanda-data/redpanda/blob/9e9db8e0ee259775f03ff9667a0310959ed31c81/tests/rptest/utils/expect_rate.py#L12-L20).

ajfabbri commented on tools/cloud_storage_manifest_patcher/main.py:
```diff
@@ -0,0 +1,393 @@
+        self._region = region
+        self.logger = logger
+
+    def create_bucket(self, name):
```
Thank you for adding some type annotations. Ideally we'd have all function types (args and return values) annotated as outlined in our (evolving) python standards (see wiki page "Python Standardization" for details. I'm happy to help add annotations later (as followup PR) if that helps.

ajfabbri commented on tools/cloud_storage_manifest_patcher/main.py:
```diff
@@ -309,46 +370,103 @@ def list_buckets(self) -> dict[str, Union[list, dict]]:
+        total_size += batch.batch_size
+        last_offset = max(last_offset, batch.base_offset + batch.record_count)
+
+    return base_offset, base_timestamp, last_offset, max_timestamp, num_configs, total_size
```
Seems like a decent case for `typing.NamedTuple`. (This can be done as followup PR if it helps.)

ajfabbri commented on tools/cloud_storage_manifest_patcher/main.py:
```diff
@@ -357,11 +475,15 @@ def generate_options():
+            segments.append(meta.Key)
+
+    print(
+        f"Scanned the bucket, {num_objects} objects scanned, {segments_size} bytes in {segments_num} segments, {len(topic_manifests)} topics found, {len(partition_manifests)} partitions found"
```
nit: line length.

ajfabbri commented on tools/cloud_storage_manifest_patcher/main.py:
```diff
@@ -376,18 +498,73 @@ def generate_options():
+                    "topic": topic,
+                    "partition": part_id,
+                    "revision": revision,
+                    "last_offset": 0,  # TODO: update
```
TODO is already done below, right?

ajfabbri commented on src/v/cloud_storage/tests/topic_manifest_test.cc:
```diff
@@ -69,6 +69,24 @@ static constexpr std::string_view full_topic_manifest_json = R"json({
 
+static constexpr std::string_view topic_manifest_json_with_prev_revision
+  = R"json({
+    "version": 1,
```
Is there any need to bump this version? I assume new code treats the new `prev_revision_id` as an optional; it will not fail it is is missing. What will old code do if it reads a manifest with this new field?

ajfabbri commented on src/v/cloud_storage/partition_recovery_manager.cc:
```diff
@@ -543,7 +543,7 @@ partition_downloader::find_recovery_material(const remote_manifest_path& key) {
         throw missing_partition_exception(key);
     }
-    auto orig_rev = recovery_mat.topic_manifest.get_revision();
+    auto orig_rev = recovery_mat.topic_manifest.get_prev_revision();
```
Do we ever need to clear `prev_revision`? Trying to understand what happens when partitions finish their recovery, and then something happens and we need to recover again.

ajfabbri commented:
Made a pass over this. Overall I think this looks like a good improvement. I don't fully understand all the details of the recovery algorithm, but I've asked some "breaker mindset" questions that may help. What sort of automated test coverage do we have for recovery? Do we have PRs filed for any followup test development that we need?

Lazin reviewed (commented):
Will iterate on it one more time I guess

Lazin commented:
cc @ajfabbri and @dotnwat 

I reworked the C++ part of the PR. Instead of adding a field to the topic manifest the I changed the workflow. Now the topic manifest is downloaded before recovery starts once for all replicas/partitions. This eliminates race condition.

Lazin commented:
I removed the python script from this PR since it changes every day as we work on an incident. I'll add a new PR with it and address all the comments related to the script in that PR.

Lazin commented:
It depends on flags.You can run recovery on a topic with disabled remote-read/remote-write.

LenaAn commented:
talked offline with @Lazin , we agreed that
1. shadow indexing read path is already using initial revision id from `remote_topic_properties`, see `topic_table.cc:51`, so it should be possible not to upload partition manifest with another initial revision id during topic recovery (in `partition_downloader::download_log(const remote_manifest_path& manifest_key)`)
2. This will make the `partition_recovery_manager` logic simpler, access to S3 read-only during topic recovery even with archival and shadow indexing enabled, and it will be possible to recover topic multiple times from the same bucket (since topic recovery will not have side effects )

Lazin commented:
cc @LenaAn 

I implemented the proposed change. Looks like the tests are passing. Thanks for the tip!

LenaAn commented on src/v/cloud_storage/partition_recovery_manager.cc:
```diff
@@ -252,24 +252,19 @@ partition_downloader::download_log(const remote_manifest_path& manifest_key) {
     }
     if (cst_log.is_enabled(ss::log_level::debug)) {
         std::stringstream ostr;
-        target.serialize(ostr);
```
looks like `target` is not used anymore so we can get rid of it.

LenaAn commented on src/v/cluster/remote_topic_configuration_source.cc:
```diff
@@ -49,7 +50,65 @@ remote_topic_configuration_source::set_remote_properties_in_config(
+}
+
+ss::future<errc>
+remote_topic_configuration_source::set_recovered_topic_properties(
```
looks like this method does the same as set_remote_properties_in_config +  sets partition_count and calls `apply_retention_defaults`. Maybe they can share some logic?

Lazin replied:
I already extracted `download_topic_manifest` method.

LenaAn commented:
And I think the last commit can be merged with the first one since you delete all the partition manifest upload logic.

Cool changes!

LenaAn commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -457,6 +460,46 @@ ss::future<topic_result> topics_frontend::do_create_topic(
               ->remote_partition_count;
     }
 
+    if (assignable_config.is_recovery_enabled()) {
```
nit: maybe add an assert that `is_read_replica` and `is_recovery_enabled` are not true at the same time? Cause otherwise `remote_topic_properties` from recovery will overwrite `remote_topic_properties` from read replica

Lazin replied:
fixed this in a different way by validating the flags in the kafka layer

LenaAn reviewed (approved):
LGTM

LenaAn commented on src/v/cloud_storage/partition_recovery_manager.cc:
```diff
@@ -395,6 +400,11 @@ partition_downloader::download_log_with_capped_size(
           }
       });
     update_downloaded_offsets(std::move(dloffsets), dlpart);
+    if (dlpart.num_files == 0) {
```
what does `num_files` mean in `download_part`?

Lazin replied:
number of files in the `*_part` directory

LenaAn reviewed (approved):
LGTM
